### PR TITLE
Minor: comments that explain the schema used in simply_expressions

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/pull/7699

## Rationale for this change

I am somewhat embarrassed that I did not fully understand @Dandandan 's fix in https://github.com/apache/arrow-datafusion/pull/7699/files and then spent some non trivial time tracking the same issue down in IOx and filed a ticket https://github.com/apache/arrow-datafusion/issues/7746 only to find it had already been fixed

## What changes are included in this PR?
Add some comments to try and explain the change more generally

BTW I made a diagram explaining what is going on, that I wanted to share even though I am not sure it belongs in the comments directly:

```
 Query:                                                                                                                 
 SELECT t.a FROM t WHERE t.b > 5                                                                                        
                                                                                                                        
                                                                                                                        
                                                                                                                        
┌────────────────────┐                                                                                                  
│  Projection(t.a)   │                                                                                                  
└────────────────────┘                                                                                                  
           ▲                                                                                                            
           │                                                                                                            
           │                                                                                                            
┌────────────────────┐        Filter           ┌────────────────────┐     Projection     ┌────────────────────┐         
│  Filter(t.b > 5)   │   ────Pushdown──▶       │  Projection(t.a)   │  ────Pushdown──▶   │  Projection(t.a)   │         
└────────────────────┘                         └────────────────────┘                    └────────────────────┘         
           ▲                                              ▲                                         ▲                   
           │                                              │                                         │                   
           │                                              │                                         │                   
┌────────────────────┐                         ┌────────────────────┐                    ┌────────────────────┐         
│        Scan        │                         │        Scan        │                    │        Scan        │         
└────────────────────┘                         │  filter=(t.b > 5)  │                    │  filter=(t.b > 5)  │         
                                               └────────────────────┘                    │  projection=(t.a)  │         
                                                                                         └────────────────────┘         
                                                                                                                        
Initial Plan                                                                           Projection pushdown notes that   
                                           If `TableProviderFilterPushDown`            the scan only needs t.a          
                                           returns true, filter pushdown                                                
                                           pushes the filter into the scan             BUT internally evaluating the    
                                                                                       predicate still requires t.b     
                                                                                                                        
```

## Are these changes tested?

<!--
We typically require tests for all PRs in order to:
1. Prevent the code from being accidentally broken by subsequent changes
2. Serve as another way to document the expected behavior of the code

If tests are not included in your PR, please explain why (for example, are they covered by existing tests)?
-->

## Are there any user-facing changes?

<!--
If there are user-facing changes then we may require documentation to be updated before approving the PR.
-->

<!--
If there are any breaking changes to public APIs, please add the `api change` label.
-->

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs [hunk 1]
```diff
@@ -66,8 +66,16 @@ impl SimplifyExpressions {
         let schema = if !plan.inputs().is_empty() {
             DFSchemaRef::new(merge_schema(plan.inputs()))
         } else if let LogicalPlan::TableScan(scan) = plan {
-            // When predicates are pushed into a table scan, there needs to be
-            // a schema to resolve the fields against.
+            // When predicates are pushed into a table scan, there is no input
+            // schema to resolve predicates against, so it must be handled specially
+            //
+            // Note that this is not `plan.schema()` which is the *output*
+            // shema, and reflects any pushed down projection. The output schema
```
Comment (comphead) [inline datafusion/optimizer/src/simplify_expressions/simplify_exprs.rs:73]:
```suggestion
            // schema, and reflects any pushed down projection. The output schema
```

**Comment by comphead** [comment]
The diagram is awesome, surely we can find a place for it in docs

**Comment by alamb** [comment]
> The diagram is awesome, surely we can find a place for it in docs

I have included it in https://github.com/apache/arrow-datafusion/pull/7759

**Comment by alamb** [comment]
@Dandandan  or @comphead  could I ask one of you to approve this PR (it needs a review prior to being able to merge):

Screenshot 2023-10-10 at 12 18 44 PM

**Review by comphead — APPROVED** [review]
lgtm

**Comment by alamb** [comment]
Thank you @comphead

**Pull request merged**